### ISSUE TITLE
MAINT: Fix undefined behaviour issues with memcpy

### ISSUE THREAD

**Issue by seberg** [numpy/numpy#13654]
Followup to gh-13590. `strides` and `shape`/`PyArray_DIMS`, can be NULL, in which case we sometimes call `memcopy` with a `NULL` argument and copying 0 bytes. This is undefined and should be avoided.

```
git grep -E "memcpy\((PyArray_(STRIDES|DIMS)|([^-]*->)?(dim|stri))"
```
may find some such issues,
```
memcpy\(.*?, sizeof(npt_intp).*?\)
```
should find more (although I think it requires multi-line search).

It may be good to implement a `npy_memcpy` or similar helper function.

**Comment by eric-wieser** [comment]
> strides and shape/PyArray_DIMS, can be NULL

Is this documented as allowed?

**Comment by seberg** [comment]
Nope, do not think so. I guess we should add the documentation label as well ;).

**Comment by mattip** [comment]
Removed Documentation label since that part was fixed

**Comment by tapaswenipathak** [comment]
Hi folks: Interested, can I work on the ticket?

**Comment by rgommers** [comment]
@tapaswenipathak please go for it. a general note, if it's clear what to do and there are no linked PRs or anyone saying they're working on it, you can just go ahead. we don't really work with assigning issues, anyone is free to pick up any issue

**Comment by tapaswenipathak** [comment]
Sure, ack; thanks @rgommers.
